library/hatomap.py

```diff
@@ -0,0 +1,478 @@
+            ).belongs_tile(),
+            WebMercatorPixelCoord(
+                self.pixel_x_east, self.pixel_y_south, self.zoom
+            ).belongs_tile(),
+        )
+
+    def geocoord2pixel(self, geocoord: GeoCoord) -> Tuple[int, int]:
+        wmp_coord = WebMercatorPixelCoord.from_geocoord(geocoord, self.zoom)
+        return (
+            int(round(wmp_coord.pixel_x - self.pixel_x_west)),
+            int(round(wmp_coord.pixel_y - self.pixel_y_north)),
+        )
+
+    def _ndarray_geocoords2pixel(self, geocoords: np.ndarray) -> np.ndarray:
+        return (
+            np.array(
+                [
+                    256 * (1 << self.zoom) * (geocoords[..., 1] + 180) / 360
+                    - self.pixel_x_west,
+                    256
+                    * (1 << self.zoom)
+                    * (
+                        0.5
+                        - np.log(
+                            np.tan(np.pi / 4 + geocoords[..., 0] * np.pi / 180 / 2)
+                        )
+                        / (2 * np.pi)
+                    )
+                    - self.pixel_y_north,
+                ]
+            )
+            .astype(np.int32)
+            .T
+        )
+
+    def geocoords2pixel(
+        self,
+        geocoords: Union[
+            np.ndarray, List[np.ndarray], List[GeoCoord], List[List[GeoCoord]]
+        ],
+    ) -> Any:
+
+        if type(geocoords) is np.ndarray:
+            return self._ndarray_geocoords2pixel(geocoords)
+        if type(geocoords) is not list:
+            return None
+        if type(geocoords[0]) is np.ndarray:
+            return [self._ndarray_geocoords2pixel(g) for g in geocoords]
+
+        if type(geocoords[0]) is GeoCoord:
+            return [self.geocoord2pixel(g) for g in geocoords]
+        if type(geocoords[0]) is list and type(geocoords[0][0]) is GeoCoord:
+            return [[self.geocoord2pixel(g) for g in gg] for gg in geocoords]
+
+
+@dataclass
+class MapBox:
+    center: GeoCoord = GeoCoord(5, 40)
+    zoom: int = 3
+    width: int = 800
+    height: int = 800
+
+    def pixelbbox(self) -> WebMercatorPixelBBox:
+        center_pixel = WebMercatorPixelCoord.from_geocoord(self.center, self.zoom)
+        origin_pixel_x = int(round(center_pixel.pixel_x - self.width / 2))
+        origin_pixel_y = int(round(center_pixel.pixel_y - self.height / 2))
+        return WebMercatorPixelBBox(
+            pixel_x_west=origin_pixel_x,
+            pixel_x_east=origin_pixel_x + self.width,
+            pixel_y_north=origin_pixel_y,
+            pixel_y_south=origin_pixel_y + self.height,
+            zoom=self.zoom,
+        )
+
+
+@dataclass
+class RasterTileServer:
+    url: str
+
+    @staticmethod
+    def _get_image_content(url):
+        return cv2.imdecode(
+            np.asarray(bytearray(requests.get(url).content), dtype=np.uint8), -1
+        )
+
+    def request(self, bbox: WebMercatorPixelBBox = None) -> np.ndarray:
```

Comment:
```suggestion
    def request(self, bbox: WebMercatorPixelBBox = None) -> Optional[np.ndarray]:
```

https://github.com/dev-hato/hato-bot/actions/runs/2783525807/jobs/4380588635

```
library/hatomap.py:185: error: Incompatible return value type (got "None", expected "ndarray[Any, Any]")
```

上記が正しそう。